docs/proposals/pod-lifecycle-event-generator.md

```diff
@@ -0,0 +1,234 @@
+shields Kubelet from the runtime specifics.
+
+```go
+type PodLifeCycleEventType string
+
+const (
+    ContainerStarted      PodLifeCycleEventType = "ContainerStarted"
+    ContainerStopped      PodLifeCycleEventType = "ContainerStopped"
+    NetworkSetupCompleted PodLifeCycleEventType = "NetworkSetupCompleted"
+    NetworkFailed         PodLifeCycleEventType = "NetworkFailed"
+)
+
+// PodLifecycleEvent is an event reflects the change of the pod state.
+type PodLifecycleEvent struct {
+    // The pod ID.
+    ID types.UID
+    // The type of the event.
+    Type PodLifeCycleEventType
+    // The accompanied data which varies based on the event type.
+    Data interface{}
+}
+```
+
+Using Docker as an example, starting of a POD infra container would be
+translated to a NetworkSetupCompleted`pod lifecycle event.
+
+
+## Pod Lifecycle Event Generator (PLEG)
+
+PLEG watches the upstream container events, and translates the events to
+corresponidng pod lifecycle events. The upstream container events could come
+from:
+
+(1). *Event stream provided by each container runtime*
+
+Docker's API exposes an [event
+stream](https://docs.docker.com/reference/api/docker_remote_api_v1.17/#monitor-docker-s-events).
+Nonethelss, rkt does not support this yet, but they will eventually support it
+(see [coreos/rkt#1193](https://github.com/coreos/rkt/issues/1193)).
+
+(2). *cgroups event stream by cAdvisor*
+
+cAdvisor is integrated in Kubelet to provide container stats. It watches cgroups
+containers using inotify and exposes an event stream. Even though it does not
+support rkt yet, it should be straightforward to add such a support.
+
+Option (1) may provide richer sets of events, but option (2) has the advantage
+to be more universal across runtimes, as long as the container runtime uses
+cgroups. Regardless of what one chooses to implement now, the container event
+stream should be easily swappable with a clearly defined interface.
+
+### Relist
+
+PLEG also needs to relist (e.g., `docker ps`) periodically to check if
+there are any missing events. We can use periodic relisting as a
+fallback option if there is no container event stream for a specific
+runtime. Although this is similar to Kubelet's polling today, it will
+only be performed by a single thread (PLEG).  This means that we still
+benefit from not having all pod workers hitting the container runtime
+concurrently. Moreover, only the relevant pod worker would be woken up
+to perform a sync.
+
+Note that there are some caveats in the implementation details. The
+container events and the output of relisting command may not provide
+all necessary information to generate a pod lifecycle event, PLEG
+would need to examine (e.g., `docker inspect`) the container for
+details.  It would also need to filter out outdated events occured
+before the relisting to avoid duplicated events.
+
+### Bootstrapping/restart
+
+Upon restart, PLEG needs to perform a reslit to retrieve all container
```

Comment:
Docker restart is another case we have to detect and handle here too. You need to detect event_stream is disconnect. When docker is starting up, it might send you tons of events of old containers.